src/mrsimulator/utils/collection.py

```diff
@@ -220,100 +225,109 @@ def site_generator(
-            attributes.append(obj_ext)
-        else:
-            attributes.append(None)
-
-    # Attributes order is same as below in list comprehension
-    attributes = [_extend_to_nparray(attr, n_sites) for attr in attributes]
-
-    return np.asarray(
-        [
-            Site(
-                isotope=iso,
-                isotropic_chemical_shift=shift,
-                name=name,
-                label=label,
-                description=desc,
-                shielding_symmetric=symm,
-                shielding_antisymmetric=antisymm,
-                quadrupolar=quad,
-            )
-            for iso, shift, name, label, desc, symm, antisymm, quad in zip(*attributes)
-        ]
-    )
-
-
-def _fix_item(item):
-    """Flattens multidimensional arrays into 1d array."""
-    if isinstance(item, (list, np.ndarray)):
-        return np.asarray(item).ravel()
-    return item
-
-
-def _extend_to_nparray(item, n):
-    """If item is already list/array return np.array, otherwise extend to length n."""
-    data = item if isinstance(item, (list, np.ndarray)) else [item for _ in range(n)]
-    return np.asarray(data)
-
-
-def _extend_dict_values(_dict, n_sites):
-    """Checks and extends dict values. Returns dict or list of dicts and max length."""
-    _dict = {key: _fix_item(val) for key, val in _dict.items()}
-    n_sites_dict = _check_lengths(list(_dict.values()))
-    if n_sites != 1 and n_sites_dict != 1 and n_sites != n_sites_dict:
-        raise ValueError(f"A list in a dictionary was misshapen. {LIST_LEN_ERROR_MSG}")
-
-    if n_sites_dict == 1:
-        _dict = {
-            key: val[0] if isinstance(val, (list, np.ndarray)) else val
-            for key, val in _dict.items()
-        }
-        return _dict, 1
-
-    _dict = {key: _extend_to_nparray(val, n_sites_dict) for key, val in _dict.items()}
-    return _zip_dict(_dict), n_sites_dict
+    isotope = _flatten_item(isotope)
+    isotropic_chemical_shift = _flatten_item(isotropic_chemical_shift)
+    shielding_symmetric = _flatten_item(shielding_symmetric)
+    shielding_antisymmetric = _flatten_item(shielding_antisymmetric)
+    quadrupolar = _flatten_item(quadrupolar)
+    name = _flatten_item(name)
+    label = _flatten_item(label)
+    description = _flatten_item(description)
+    args = [
+        isotope,
+        isotropic_chemical_shift,
+        shielding_symmetric,
+        shielding_antisymmetric,
+        quadrupolar,
+        name,
+        label,
+        description,
+    ]
 
+    n_sites = _check_lengths_of_args(*args)
 
-def _check_lengths(attributes):
-    """Ensures all attribute lengths are 1 or maximum attribute length."""
-    lengths = np.array([np.asarray(attr).size for attr in attributes])
+    return list(_site_generator(n_sites, *args))
 
-    if np.all(lengths == 1):
-        return 1
 
-    lengths = lengths[np.where(lengths != 1)]
-    if np.unique(lengths).size == 1:
-        return lengths[0]
+def _site_generator(  # noqa: C901
```

Comment:
What is the reason for noqa: C901? I do not see any warnings from the linter.

Reply:
Should have ignored method to complex error for CodeFactor. Worked locally when running pre-commit, but does not seem to pass in the PR

Reply:
CodeFactor checks are warning, not required for the PR to merge. Ignore it for now.